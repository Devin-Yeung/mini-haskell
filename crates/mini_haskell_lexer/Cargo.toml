[package]
name = "mini_haskell_lexer"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
logos = "0.13.0"
insta = { workspace = true, features = ["glob"] }
thiserror = { workspace = true }
mini_haskell_diagnostic = { path = "../mini_haskell_diagnostic" }
testsuite = { path = "../testsuite" }
testsuite_macros = { path = "../testsuite_macros" }
