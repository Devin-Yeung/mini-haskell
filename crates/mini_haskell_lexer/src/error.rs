use mini_haskell_diagnostic::diagnostic::Diagnostic;
use mini_haskell_diagnostic::span::Span;
use thiserror::Error;

#[derive(Error, Debug, PartialEq, Clone)]
#[non_exhaustive]
pub enum LexingError {
    #[error("Unexpected token")]
    UnexpectedToken(Span),
}

impl Diagnostic for LexingError {
    fn span(&self) -> Span {
        match self {
            LexingError::UnexpectedToken(span) => span.clone(),
        }
    }

    fn message(&self) -> String {
        self.to_string()
    }
}
