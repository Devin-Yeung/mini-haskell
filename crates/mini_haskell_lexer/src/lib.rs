mod diagnostic;
pub mod error;
pub mod lexer;
