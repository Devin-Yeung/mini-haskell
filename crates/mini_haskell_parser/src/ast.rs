use mini_haskell_diagnostic::span::Span;

#[derive(Debug)]
pub struct Expr {
    kind: ExprKind,
    span: Span,
}

#[derive(Debug)]
pub enum ExprKind {
    Literal(Literal),
    BinaryExpr(BinaryExpr),
    CondExpr(CondExpr),
}

#[derive(Debug)]
pub enum Literal {
    NatureNum(usize),
    Bool(bool),
}

#[derive(Debug)]
pub enum BinaryOp {
    Plus,
    Ampersand,
    Less,
    Equal,
}

#[derive(Debug)]
pub struct BinaryExpr {
    lhs: Box<Expr>,
    op: BinaryOp,
    rhs: Box<Expr>,
}

#[derive(Debug)]
pub struct CondExpr {
    condition: Box<Expr>,
    then_branch: Box<Expr>,
    else_branch: Box<Expr>,
}

impl Expr {
    pub fn new(kind: ExprKind, span: Span) -> Self {
        Self { kind, span }
    }
}
