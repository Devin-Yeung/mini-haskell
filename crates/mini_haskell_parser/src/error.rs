use miette::Diagnostic;
use mini_haskell_diagnostic::span::Span;
use thiserror::Error;

#[derive(Diagnostic, Error, Debug, Eq, PartialEq, Copy, Clone)]
pub enum SyntaxError {
    #[error("Unexpected EOF")]
    UnexpectedEOF,
    #[error("SyntaxError: Unexpected token")]
    UnexpectedToken {
        #[label = "Expected {expected}, found {found}"]
        span: Span,
        expected: &'static str,
        found: &'static str,
    },
    #[error("Expected {0}")]
    Expected(&'static str),
}
