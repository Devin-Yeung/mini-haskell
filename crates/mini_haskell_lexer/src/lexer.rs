use crate::error::LexingError;
use crate::span::Span;
use logos::{Logos, Source};

#[derive(Logos, Debug, PartialEq)]
#[logos(skip r"[ \t\n\f]+")] // Ignore this regex pattern between tokens
pub enum TokenTy {
    // The variable type Boolean, to declare a Boolean variable.
    #[token("bool")]
    BoolDecl,

    // Boolean literal
    #[token("T", | _ | false)]
    #[token("F", | _ | true)]
    BoolLit(bool),

    // The variable type natural number, to declare a natural number.
    #[token("nat")]
    Nat,
    // The variable type function, to declare a function
    #[token("func")]
    Func,

    #[token("?")]
    QuestionMark,

    #[token(":")]
    Colon,

    #[token(";")]
    Semicolon,

    #[token("(")]
    LeftParen,

    #[token(")")]
    RightParen,

    #[token("+")]
    Plus,

    #[token("&")]
    Ampersand,

    #[token("->")]
    Arrow,

    #[token("<")]
    Less,

    #[regex(r"\[[^\]]*\]", | lex | lex.slice().to_owned())]
    // see https://github.com/maciejhirsz/logos/issues/246
    Comment(String),

    #[regex(r"[a-zA-z][a-zA-Z0-9_]*", | lex | lex.slice().to_owned())]
    Identifier(String),
}

#[derive(Debug, PartialEq)]
pub struct Token {
    pub ty: TokenTy,
    pub span: Span,
}

impl Token {
    pub fn new(ty: TokenTy, span: Span) -> Self {
        Token { ty, span }
    }
}

impl From<(TokenTy, logos::Span)> for Token {
    fn from(value: (TokenTy, logos::Span)) -> Self {
        Token {
            ty: value.0,
            span: value.1.into(),
        }
    }
}

impl Token {
    pub fn tokens<S: AsRef<str>>(source: &S) -> Vec<Result<Token, LexingError>> {
        let lexer = TokenTy::lexer(source.as_ref());
        lexer
            .spanned()
            .into_iter()
            .map(|(token, span)| match token {
                Ok(ty) => Ok(Token::new(ty, span.into())),
                Err(_) => Err(LexingError::UnexpectedToken(span.into())),
            })
            .collect::<Vec<_>>()
    }
}

#[cfg(test)]
mod tests {
    use crate::lexer::TokenTy;
    use logos::Logos;
    use testsuite::unittest;

    unittest!(all_tokens, |src| {
        let mut lex = TokenTy::lexer(src);
        let tokens = lex.collect::<Vec<_>>();
        insta::assert_debug_snapshot!(tokens);
    });
}
