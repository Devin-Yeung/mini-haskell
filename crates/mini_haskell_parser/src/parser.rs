use crate::ast::{Expr, Literal};
use crate::error::SyntaxError;
use mini_haskell_lexer::lexer::{TokenTy, Tokenizer};
use std::iter::Peekable;

pub struct Parser<'src> {
    tokenizer: Peekable<Tokenizer<'src>>,
}

impl<'src> Parser<'src> {
    pub fn new<S: AsRef<str> + ?Sized>(src: &'src S) -> Self {
        Self {
            tokenizer: Tokenizer::new(src).peekable(),
        }
    }

    fn consume(&mut self, ty: TokenTy) -> Result<TokenTy, SyntaxError> {
        dbg!(self.tokenizer.peek());
        match self.peek_type()? {
            found if found == ty => Ok(self.advance()?),
            found => Err(SyntaxError::UnexpectedToken {
                expected: ty.name(),
                found: found.name(),
            }),
        }
    }

    fn consume_if(&mut self, ty: TokenTy) -> bool {
        if self.peek_type() == Ok(ty) {
            self.advance().unwrap();
            return true;
        }
        false
    }

    fn peek_type(&mut self) -> Result<TokenTy, SyntaxError> {
        dbg!(self.tokenizer.peek());
        match self.tokenizer.peek() {
            None => Ok(TokenTy::EOF),
            Some(Ok(token)) => Ok(token.ty.clone()),
            Some(Err(_)) => self.advance(),
        }
    }

    fn advance(&mut self) -> Result<TokenTy, SyntaxError> {
        self.tokenizer
            .next()
            .map_or(Err(SyntaxError::UnexpectedEOF), |inner| {
                match inner {
                    Ok(tok) => Ok(tok.ty),
                    Err(_) => {
                        // lexing error is ignored and skip to next valid token
                        loop {
                            match self.tokenizer.peek() {
                                None => return Err(SyntaxError::UnexpectedEOF),
                                Some(Ok(tok)) => return Ok(tok.ty.clone()),
                                Some(Err(_)) => {
                                    self.tokenizer.next();
                                }
                            }
                        }
                    }
                }
            })
    }

    /// parse primary expression according to following rules:
    ///
    /// ```text
    /// primary  â†’ NAT | "T" | "F" | IDENTIFIER ;
    /// ```
    fn primary(&mut self) -> Result<Expr, SyntaxError> {
        match self.peek_type()? {
            TokenTy::BoolLit(b) => Ok(Expr::Literal(Literal::Bool(b))),
            TokenTy::NatLit(n) => Ok(Expr::Literal(Literal::NatureNum(n))),
            TokenTy::Identifier(_) => todo!(),
            _ => Err(SyntaxError::Expected("expression")),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::parser::Parser;
    use crate::parser::SyntaxError;
    use crate::parser::TokenTy;
    use mini_haskell_lexer::lexer::TokenTy::NatLit;
    use testsuite::unittest;

    unittest!(advance, |src| {
        let mut parser = Parser::new(src);
        let mut result = Vec::<Result<TokenTy, SyntaxError>>::new();
        loop {
            match parser.advance() {
                Err(SyntaxError::UnexpectedEOF) => break,
                Ok(tok) => result.push(Ok(tok)),
                Err(err) => result.push(Err(err)),
            }
        }
        insta::assert_debug_snapshot!(result);
    });

    unittest!(consume, |src| {
        let mut parser = Parser::new(src);
        let result = vec![
            parser.peek_type(),
            parser.consume(NatLit(1)),
            parser.peek_type(),
            parser.consume(NatLit(2)),
            parser.peek_type(),
        ];
        insta::assert_debug_snapshot!(result);
    });

    unittest!(primary, |src| {
        let asts = src
            .split('\n')
            .map(|line| Parser::new(line).primary())
            .collect::<Vec<_>>();
        insta::assert_debug_snapshot!(asts);
    });
}
