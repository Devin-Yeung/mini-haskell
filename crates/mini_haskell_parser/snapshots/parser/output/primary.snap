---
source: crates/mini_haskell_parser/src/parser.rs
expression: asts
---
[
    Ok(
        Expr {
            kind: Literal(
                Bool(
                    false,
                ),
            ),
            span: Span {
                start: 0,
                end: 1,
            },
        },
    ),
    Ok(
        Expr {
            kind: Literal(
                Bool(
                    true,
                ),
            ),
            span: Span {
                start: 0,
                end: 1,
            },
        },
    ),
    Ok(
        Expr {
            kind: Literal(
                NatureNum(
                    123,
                ),
            ),
            span: Span {
                start: 0,
                end: 3,
            },
        },
    ),
]
